pyinjective/core/network.py

```diff
@@ -1,171 +1,103 @@
+        return False
 
-        return cookie
 
-    async def exchange_cookie(self, metadata_query_provider: Callable) -> str:
-        if self._exchange_cookie is None:
-            async with self._exchange_cookie_initialization_lock:
-                if self._exchange_cookie is None:
-                    await self._fetch_exchange_cookie(metadata_query_provider=metadata_query_provider)
-        cookie = self._exchange_cookie
-        self._check_exchange_cookie_expiration()
+class ExpiringCookieAssistant(CookieAssistant):
+    def __init__(self, expiration_time_keys_sequence: List[str], time_format: str):
+        self._cookie: Optional[str] = None
+        self._expiration_time_keys_sequence = expiration_time_keys_sequence
+        self._time_format = time_format
 
-        return cookie
+    @classmethod
+    def for_kubernetes_public_server(cls):
+        return cls(
+            expiration_time_keys_sequence=["grpc-cookie", "expires"],
+            time_format="%a, %d-%b-%Y %H:%M:%S %Z",
+        )
 
-    async def _fetch_chain_cookie(self, metadata_query_provider: Callable):
-        metadata = await metadata_query_provider()
-        cookie_info = next((value for key, value in metadata if key == "set-cookie"), None)
+    def cookie(self) -> Optional[str]:
+        self._check_cookie_expiration()
 
-        if cookie_info is None:
-            raise RuntimeError(f"Error fetching chain cookie ({metadata})")
+        return self._cookie
 
-        self._chain_cookie = cookie_info
+    async def process_response_metadata(self, grpc_call: Call):
+        metadata = await grpc_call.initial_metadata()
+        if "set-cookie" in metadata:
+            self._cookie = metadata["set-cookie"]
 
-    async def _fetch_exchange_cookie(self, metadata_query_provider: Callable):
-        metadata = await metadata_query_provider()
-        cookie_info = next((value for key, value in metadata if key == "set-cookie"), None)
+    def _check_cookie_expiration(self):
+        if self._is_cookie_expired():
+            self._cookie = None
 
-        if cookie_info is None:
-            cookie_info = ""
+    def _is_cookie_expired(self) -> bool:
+        is_expired = False
 
-        self._exchange_cookie = cookie_info
+        if self._cookie is not None:
+            cookie = SimpleCookie()
+            cookie.load(self._cookie)
+            cookie_map = cookie
 
-    def _check_chain_cookie_expiration(self):
-        if self._is_cookie_expired(cookie_data=self._chain_cookie):
-            self._chain_cookie = None
+            for key in self._expiration_time_keys_sequence[:-1]:
+                cookie_map = cookie.get(key, {})
 
-    def _check_exchange_cookie_expiration(self):
-        if self._is_cookie_expired(cookie_data=self._exchange_cookie):
-            self._exchange_cookie = None
+            expiration_data: Optional[str] = cookie_map.get(self._expiration_time_keys_sequence[-1], None)
+            if expiration_data is not None:
```

Comment:
So if key is missing in the cookie_map then it's not expired. Is this confirmed to be this way by default?

Reply:
Yes, correct. I have tested it with the current bare metal load balanced servers, and the k8s load balanced servers. If cookie is not renewed, then the response from the server does not include the info